### ISSUE TITLE
Primitive polynomial over Galois fields.

### ISSUE THREAD

**Issue by sylee957** [sympy/sympy#18693]
Before any proper work is done for finite field over `p**n`, I couldn't find any work done for primitive polynomial in `galoistools`.
But I think that it should be done first, because I guess the only computationally constructive way to develop an arithmetic over finite field is by finding a primitive polynomial.

```python3
from sympy.polys.galoistools import gf_irreducible_p, gf_trunc, gf_rem, gf_degree

def gf_primitive_p(f, p, K):
    if not gf_irreducible_p(f, p, K):
        return False

    f = gf_trunc(f, p)
    n = gf_degree(f)
    for l in range(n, p**n-1):
        c = [1] + [0]*(l-1) + [p-1]
        rem = gf_rem(c, f, p, K)
        if not rem:
            return False
    return True
```

I could only find one improvement over the naive algorithm in 
https://www.researchgate.net/publication/4137376_Fast_identification_of_primitive_polynomials_over_Galois_fields_results_from_a_course_project
as I had followed the references in MATLAB
So that can be used instead of this naive construction.

I would also expect this to be a public method of `Poly`, but since `is_primitive` is conflicting, I guess a different name should be used.
I just wonder if this can be a valid addition to `galoistools`.

**Comment by jksuom** [comment]
I think that prime power fields can be constructed with all irreducible polynomials. Do you know of any computational advantage that a primitive polynomial would give?

**Comment by sylee957** [comment]
Okay, I'd double check that all the irreducible polynomial can build prime power fields.
Though it can just be a different question of adding the predicate or not.

But I also wonder the computational advantage that using primitive polynomial can provide over irreducible polynomial.
e.g. https://math.stackexchange.com/tags/finite-fields/info
> Primitive polynomials are extremely useful in software implementations of the arithmetic of a moderate size finite field. This is largely because having a primitive polynomial at hand allows one to generate look-up-tables for both the base g discrete logarithm as well as its inverse function. See this CW question for examples.
> For that reason extensive tables of primitive polynomials have been generated. One such table is here.

**Comment by sylee957** [comment]
At least it looks like checking if the polynomial is primitive, can simultaneously reveal the multiplication table if it is done like the posted link.

**Comment by gschintgen** [comment]
I agree with the OP that it should be possible to request the instantiation of `GF(p**n)` with a primitive polynomial. The reason being ease of use (immediate availability of a generator of the cyclic multiplicative group). The API should allow for e.g. `GF(mod, degree, primitive=True)` or even `GF(...., method='somemethod')` for choosing how the irreducible polynomial is determined.

I've seen similar ideas in the docs of both GAP and Sagemath. Both also mention a database of Conway polynomials. So that could be a complement to the naive method. But I don't think that it's truly necessary to add the function of the OP strictly before implementing `GF(p**n)`. As a first step it should be fine without as long as the user is free to pass an irreducible polynomial of their own choice.
